vortex-file/src/read/layouts/chunked.rs

```diff
@@ -131,42 +118,125 @@ impl ChunkedLayout {
             .collect::<Vec<_>>()
     }
 
-    fn child_layouts<C: Fn(LayoutPartId) -> RelativeLayoutCache>(
-        &self,
-        cache: C,
-    ) -> VortexResult<VecDeque<RangedLayoutReader>> {
+    fn children_layouts(&self) -> VortexResult<Vec<RangedLayoutReader>> {
         self.children()
-            .zip_eq(self.child_ranges())
+            .zip_eq(self.children_ranges())
             .map(|((i, c), (begin, end))| {
                 let layout = self.layout_builder.read_layout(
                     self.fb_bytes.clone(),
                     c._tab.loc(),
                     self.scan.clone(),
-                    cache(i as LayoutPartId),
+                    self.message_cache
+                        .relative(i as u16, self.message_cache.dtype().clone()),
                 )?;
                 Ok(((begin, end), layout))
             })
-            .collect::<VortexResult<VecDeque<_>>>()
+            .collect::<VortexResult<Vec<_>>>()
+    }
+
+    pub fn build(&self) -> VortexResult<ChunkedLayoutReader> {
+        Ok(ChunkedLayoutReader::new(
+            self.children_layouts()?,
+            self.metadata_layout()?,
+        ))
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct ChunkedLayoutReader {
+    layouts: Vec<RangedLayoutReader>,
+    metadata_layout: Option<Box<dyn LayoutReader>>,
+    in_progress_ranges: InProgressRanges,
+}
+
+impl ChunkedLayoutReader {
+    pub fn new(
+        layouts: Vec<RangedLayoutReader>,
+        metadata_layout: Option<Box<dyn LayoutReader>>,
+    ) -> Self {
+        Self {
+            layouts,
+            metadata_layout,
+            in_progress_ranges: RwLock::new(HashMap::new()),
+        }
+    }
+
+    fn buffer_read(&self, mask: &RowMask) -> VortexResult<Vec<MessageLocator>> {
+        let layouts_in_range = self
+            .layouts
+            .iter()
+            .filter(|((begin, end), _)| mask.end() >= *begin && mask.begin() <= *end)
+            .collect::<Vec<_>>();
```

Comment:
this needs some cache, previously we would terminate quickly since we either skipped over or took the next one